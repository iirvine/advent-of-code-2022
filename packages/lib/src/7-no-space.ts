import { readFile } from "./util";

type cd = "cd";
type ls = "ls";
type cmd = cd | ls;

const INPUT = readFile("./inputs/7-input.txt").split("\n");

interface Command {
  cmd: cmd;
  args: string[];
  output?: string[];
}

interface File {
  name: string;
  size: number;
}

interface Node {
  name: string;
  files: File[];
  directories: string[];
}

interface State {
  cwd: string;
  filetree: Record<string, Node>;
}

const Root = "/";

export function parseCommand(s: string, output?: string[]): Command {
  const [prompt, cmd, ...args] = s.split(" ");
  return {
    cmd: cmd as cmd,
    args,
    output,
  };
}

export function parent(p: string) {
  const parts = p.split("/");
  return parts.length > 2 ? parts.slice(0, parts.length - 1).join("/") : Root;
}

export function changePath(cwd: string, path: string) {
  return path === Root ? Root : cwd === Root ? `/${path}` : `${cwd}/${path}`;
}

export function buildState(input: string[]): State {
  return input.reduce<State>(
    (acc, curr) => {
      if (curr.startsWith("$")) {
        const cmd = parseCommand(curr);
        switch (cmd.cmd) {
          case "cd":
            if (cmd.args[0] == "..") {
              return { ...acc, cwd: parent(acc.cwd) };
            } else {
              return {
                ...acc,
                cwd: changePath(acc.cwd, cmd.args[0]),
              };
            }
          case "ls":
            return acc;
        }
      }
      let node = acc.filetree[acc.cwd] ?? {
        name: acc.cwd,
        files: [],
        directories: [],
      };
      const [kind, name] = curr.split(" ");
      if (kind === "dir") {
        node.directories.push(name);
      } else {
        node.files.push({ name, size: parseInt(kind) });
      }

      acc.filetree[acc.cwd] = node;
      return acc;
    },
    { cwd: "/", filetree: {} },
  );
}

export function total(n: Node, filetree: Record<string, Node>) {
  const files = n.files.reduce((acc, curr) => acc + curr.size, 0);
  const dirs = n.directories.reduce(
    (acc, curr) => acc + total(filetree[changePath(n.name, curr)], filetree),
    0,
  );
  return files + dirs;
}

export function day7_pt1() {
  const state = buildState(INPUT);
  const cutoff = 100_000;

  return Object.entries(state.filetree).reduce<number>((acc, [k, v]) => {
    const t = total(v, state.filetree);
    if (t <= cutoff) {
      return acc + t;
    }

    return acc;
  }, 0);
}

export function day7_pt2() {
  const totalSpace = 700_00_000;
  const requiredSpace = 30_000_000;
  const state = buildState(INPUT);

  const rootTotal = total(state.filetree[Root], state.filetree);
  const freeSpace = totalSpace - rootTotal;
  const needed = requiredSpace - freeSpace;

  return Object.entries(state.filetree)
    .map(([k, v]) => {
      const t = total(v, state.filetree);
      return [k, t];
    })
    .filter(([, v]) => v > needed)
    .sort(([, a], [, b]) => a - b)
    .at(0);
}
