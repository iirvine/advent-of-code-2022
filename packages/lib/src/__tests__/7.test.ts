import { buildState, day7_pt1, day7_pt2, stat } from "../7-no-space";

test("buildState", () => {
  const s = buildState([
    "$ cd /",
    "$ ls",
    "dir a",
    "dir b",
    "$ cd a",
    "$ ls",
    "1234 a.txt",
    "dir c",
    "$ cd c",
    "$ cd ..",
    "$ cd ..",
  ]);

  expect(s).toMatchSnapshot();
});

test("buildState test input", () => {
  const input = `
  $ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k
  `
    .trim()
    .split("\n");

  const s = buildState(input);
  expect(s).toMatchSnapshot();

  const t = stat(s.filetree["/a"], s.filetree);
  expect(t).toBe(29116 + 2557 + 62596 + 584);

  const t2 = stat(s.filetree["/"], s.filetree);
  expect(t2).toBe(48381165);
});

test("day7_pt1", () => {
  const s = day7_pt1();
  console.log(s);
});

test("day7_pt2", () => {
  const s = day7_pt2();
  console.log(s);
});
