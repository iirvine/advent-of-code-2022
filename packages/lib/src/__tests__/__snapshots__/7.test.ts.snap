// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildState 1`] = `
{
  "cwd": "/",
  "filetree": {
    "/": {
      "directories": [
        "a",
        "b",
      ],
      "files": [],
      "name": "/",
    },
    "/a": {
      "directories": [
        "c",
      ],
      "files": [
        {
          "name": "a.txt",
          "size": 1234,
        },
      ],
      "name": "/a",
    },
  },
}
`;

exports[`buildState test input 1`] = `
{
  "cwd": "/d",
  "filetree": {
    "/": {
      "directories": [
        "a",
        "d",
      ],
      "files": [
        {
          "name": "b.txt",
          "size": 14848514,
        },
        {
          "name": "c.dat",
          "size": 8504156,
        },
      ],
      "name": "/",
    },
    "/a": {
      "directories": [
        "e",
      ],
      "files": [
        {
          "name": "f",
          "size": 29116,
        },
        {
          "name": "g",
          "size": 2557,
        },
        {
          "name": "h.lst",
          "size": 62596,
        },
      ],
      "name": "/a",
    },
    "/a/e": {
      "directories": [],
      "files": [
        {
          "name": "i",
          "size": 584,
        },
      ],
      "name": "/a/e",
    },
    "/d": {
      "directories": [],
      "files": [
        {
          "name": "j",
          "size": 4060174,
        },
        {
          "name": "d.log",
          "size": 8033020,
        },
        {
          "name": "d.ext",
          "size": 5626152,
        },
        {
          "name": "k",
          "size": 7214296,
        },
      ],
      "name": "/d",
    },
  },
}
`;
